### ISSUE TITLE
Rename "depth-comparison" to "depth-stencil"

### ISSUE THREAD

**Issue by kainino0x** [gpuweb/gpuweb#1230]
> Talked about this with editors. Here is an expanded suggestion.
> First, we modify this bit of the spec:
> ```
>     // Texture is used with comparison sampling only.
>     "depth-comparison"
> ```
> It's already misleading for saying the textures can only be compare-sampled. I think we can instead do something like:
> ```
>     // Textures of depth/stencil formats.
>     "depth-stencil"
> ```
> So this binding type would be the only binding you could have for a texture that has a depth/stencil format. The shader compiler would therefore know precisely when it's invalid to get the 2nd, 3rd, etc component from a texture.
> 
> For stencil, we could add a separate function like `stencilLoad` returning `u32`, that would be called on those depth/stencil textures only.
> 
> I believe these measures would fully address the problem @austinEng found.

_Originally posted by @kvark in https://github.com/gpuweb/gpuweb/issues/1198#issuecomment-724322605_

**Comment by kainino0x** [comment]
Specifically this would both be more accurate and allow us to fix #1198.

**Comment by Kangz** [comment]
I don't think that `depth-stencil` is correct, because only the depth component of a depth-stencil texture can be used as `depth-comparison`. Also `depth-comparison` carries the constraint that you can only use the texture as with a `comparison-sampler`.

**Comment by kvark** [comment]
@Kangz I don't understand that answer, could you elaborate?

> only the depth component of a depth-stencil texture can be used as depth-comparison

yes, but we'll also need a way to read the stencil (without comparison)

> Also depth-comparison carries the constraint that you can only use the texture as with a comparison-sampler.

Does it? I thought we concluded that it's not the case, see https://github.com/gpuweb/gpuweb/pull/970#issuecomment-724893246

**Comment by Kangz** [comment]
> yes, but we'll also need a way to read the stencil (without comparison)

That's `GPUTextureComponentType.uint`. Actually Vulkan explicitly states that you can't use a comparison sampler with the stencil aspect of a depth-stencil texture:

> Results are undefined if:
>  - The SPIR-V instruction is one of the OpImage*Dref* instructions and the image view format is not one of the depth/stencil formats with a depth component, or the image view aspect is not VK_IMAGE_ASPECT_DEPTH_BIT.

So `depth-comparison` should at most be renamed `depth`. (I hadn't followed @kainino0x's archeology)

**Comment by kvark** [comment]
We talked about this on Matrix, and there is a way for binding the stencil aspect that I forgot about, and which @Kangz is implying. That is, binding a texture view with "stencil" aspect separately as "uint" binding. If we go this route, then the texture component type for depth doesn't need to have "-stencil" in it. I agree that it could be just "depth".

**Comment by kainino0x** [comment]
Doing so will still get us unspecified values on older Metal, so this only works if we want to insert the ,0,0,1 dynamically, or leave it unspecified, in those cases. Right?

**Comment by kvark** [comment]
We'd have to have the plumbing code to potentially mark some of the Uint textures as "stencil", internally, allowing the shader compiler to set the proper swizzle (if swizzling isn't supported by the API). it's unfortunate, but it's limited to a narrow range of platforms, and it's implementable.

**Comment by kainino0x** [comment]
If we don't know whether it's stencil in the BGL, won't we not know until draw time? So we would have to have a dynamic switch between `.rgba` and `.r001` essentially (or compile multiple pipeline variants).

We should also be sure that doing API-level swizzling actually works. It's possible that an R001 swizzle on stencil doesn't fix the unspecified result.

Also should we do this same thing for depth, so depth can be bound as float?

**Comment by kvark** [comment]
I'm not sure if we need that done for depth, that's something to consider. But we have to do something for stencil. And I agree that it would be great to first confirm if native swizzling work on those stencil views. Can Dawn tests help with this?

**Comment by kainino0x** [comment]
They certainly can. I don't know if we test it yet. @austinEng should know or be able to add those tests.

**Comment by austinEng** [comment]
>  great to first confirm if native swizzling work on those stencil views
Native swizzling on newer Metal to force to (r, 0, 0 1) ?

Note: We've also found that on some Win NVIDIA cards, (GeForce?) 7XX-series, sampling the extra components, for both depth and stencil, does not correctly return (d/s, 0, 0, 1). I've asked an NVIDIA contact to see if it's a driver issue.

**Comment by kainino0x** [comment]
> Native swizzling on newer Metal to force to (r, 0, 0 1) ?

Yes, and that it actually works on other backends like you already mentioned.

**Comment by austinEng** [comment]
On my 15-inch Macbook Pro on both Intel and AMD graphics cards, setting
```
swizzle:MTLTextureSwizzleChannelsMake(
                                                      MTLTextureSwizzleRed, MTLTextureSwizzleZero,
                                                      MTLTextureSwizzleZero, MTLTextureSwizzleOne)
```

Fixes our test

**Comment by kainino0x** [comment]
Fixed by #1223
